snapshot/storage/ql/ql.go

```diff
@@ -0,0 +1,358 @@
+package ql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"path/filepath"
+
+	"github.com/docker/containerd/snapshot"
+	"github.com/docker/containerd/snapshot/storage"
+	"github.com/pkg/errors"
+
+	_ "github.com/cznic/ql/driver"
+)
+
+type qlMetastore struct {
+	dbfile string
+}
+
+// NewMetaStore returns a snapshot MetaStore for storage of metadata related to
+// a snapshot driver backed by a ql file database. This implementation is
+// strongly consistent and does all metadata changes in a transaction to prevent
+// against process crashes causing inconsistent metadata state.
+//
+// To view mtastore with a utility use https://godoc.org/github.com/cznic/ql/ql
+func NewMetaStore(ctx context.Context, dbfile string) (storage.MetaStore, error) {
+	ms := &qlMetastore{
+		dbfile: dbfile,
+	}
+	schema := []string{
+		`CREATE TABLE IF NOT EXISTS Snapshots (Name string NOT NULL, Parent int, Active bool, Readonly bool);`,
```

Comment:
> `Parent` is an int so that it can be joined based on `id()`, which is int.

Is `id()` defined somewhere explicit or is that implicit in the schema for `ql`?

>  My immediate take is that we shouldn't need to mark deletions with a consistent backend. Having both a consistent database and data stored on disk does allow getting out of sync, making the commits multi-phase helps but a crash may still require cleaning up. I think we should focus on the database being consistent and if we have a cleanup process it would only need to cleanup leftover files on disk using the database as the source of truth.

It's not just about being consistent, but also about ensuring we have the right set of primitives to make the system repairable if something goes wrong.  This is an extremely common mitigation for disaster scenarios and is also useful for investigation if bugs get introduced.